pkg/controller/admissionchecks/multikueue/multikueuecluster.go

```diff
@@ -0,0 +1,345 @@
+	cluster := &kueuealpha.MultiKueueCluster{}
+	log := ctrl.LoggerFrom(ctx)
+
+	err := c.client.Get(ctx, req.NamespacedName, cluster)
+	if client.IgnoreNotFound(err) != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err != nil || !cluster.DeletionTimestamp.IsZero() {
+		c.stopAndRemoveCluster(req.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// get the kubeconfig
+	kubeConfig, retry, err := c.getKubeConfig(ctx, &cluster.Spec.KubeConfig)
+	if retry {
+		return reconcile.Result{}, err
+	}
+	if err != nil {
+		log.Error(err, "reading kubeconfig")
+		return reconcile.Result{}, c.updateStatus(ctx, cluster, false, "BadConfig", err.Error())
+	}
+
+	if err := c.setRemoteClientConfig(ctx, cluster.Name, kubeConfig); err != nil {
+		log.Error(err, "setting kubeconfig")
+		return reconcile.Result{}, c.updateStatus(ctx, cluster, false, "ClientConnectionFailed", err.Error())
+	}
+	return reconcile.Result{}, c.updateStatus(ctx, cluster, true, "Active", "Connected")
+}
+
+func (c *clustersReconciler) getKubeConfig(ctx context.Context, ref *kueuealpha.KubeConfig) ([]byte, bool, error) {
+	sec := corev1.Secret{}
+	secretObjKey := types.NamespacedName{
+		Namespace: c.configNamespace,
+		Name:      ref.Location,
+	}
+	err := c.client.Get(ctx, secretObjKey, &sec)
+	if err != nil {
+		return nil, !apierrors.IsNotFound(err), err
+	}
+
+	kconfigBytes, found := sec.Data[kueuealpha.MultiKueueConfigSecretKey]
+	if !found {
+		return nil, false, fmt.Errorf("key %q not found in secret %q", kueuealpha.MultiKueueConfigSecretKey, ref.Location)
+	}
+
+	return kconfigBytes, false, nil
+}
+
+func (c *clustersReconciler) updateStatus(ctx context.Context, cluster *kueuealpha.MultiKueueCluster, active bool, reason, message string) error {
+	newCondition := metav1.Condition{
+		Type:    kueuealpha.MultiKueueClusterActive,
+		Status:  metav1.ConditionFalse,
+		Reason:  reason,
+		Message: message,
+	}
+	if active {
+		newCondition.Status = metav1.ConditionTrue
+	}
+
+	// if the cluster is connected and remains that way, skip the status update
+	oldCondition := apimeta.FindStatusCondition(cluster.Status.Conditions, kueuealpha.MultiKueueClusterActive)
+	if active && oldCondition != nil && oldCondition.Status == metav1.ConditionTrue {
+		return nil
+	}
+
+	apimeta.SetStatusCondition(&cluster.Status.Conditions, newCondition)
+	return c.client.Status().Update(ctx, cluster)
+}
+
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=multikueueclusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=multikueueclusters/status,verbs=get;update;patch
+
+func newClustersReconciler(c client.Client, namespace string) *clustersReconciler {
+	return &clustersReconciler{
+		client:          c,
+		configNamespace: namespace,
+		clients:         make(map[string]*remoteClient),
+		wlUpdateCh:      make(chan event.GenericEvent, 10),
```

Comment:
Why do we create 10 channels?

Reply:
Ah, I see. We use 10 as a buffer size: https://github.com/kubernetes-sigs/kueue/blob/59980e245f202da838de890613155bf962f069bb/pkg/controller/core/core.go#L30

Thank you!